sdks/java/harness/src/main/java/org/apache/beam/fn/harness/state/CachingBeamFnStateClient.java

```diff
@@ -0,0 +1,179 @@
+
+public class CachingBeamFnStateClient implements BeamFnStateClient {
+
+  private final BeamFnStateClient beamFnStateClient;
+  private final Cache<Pair<StateKey, ByteString>, StateGetResponse> stateCache;
+  private Map<Pair<String, String>, ByteString> sideInputCacheTokens;
+  private ByteString userStateToken;
+
+  public CachingBeamFnStateClient(
+      BeamFnStateClient beamFnStateClient,
+      Cache<Pair<StateKey, ByteString>, StateGetResponse> stateCache,
+      List<CacheToken> cacheTokenList) {
+    this.beamFnStateClient = beamFnStateClient;
+    this.stateCache = stateCache;
+    this.sideInputCacheTokens = new HashMap<>();
+    this.userStateToken = null;
+
+    // Set up cache tokens
+    for (BeamFnApi.ProcessBundleRequest.CacheToken token : cacheTokenList) {
+      if (token.hasUserState()) {
+        this.userStateToken = token.getToken();
+      } else if (token.hasSideInput()) {
+        Pair<String, String> sideInput =
+            new ImmutablePair<>(
+                token.getSideInput().getTransformId(), token.getSideInput().getSideInputId());
+        sideInputCacheTokens.put(sideInput, token.getToken());
+      }
+    }
+  }
+
+  @Override
+  public void handle(
+      BeamFnApi.StateRequest.Builder requestBuilder,
+      CompletableFuture<BeamFnApi.StateResponse> response) {
+
+    StateRequest request = requestBuilder.build();
+    StateKey stateKey = request.getStateKey();
+    ByteString cacheToken = getCacheToken(stateKey);
+    StateResponse.Builder responseBuilder = StateResponse.newBuilder();
+
+    // If not cacheable proceed as normal
+    if (cacheToken == null) {
+      beamFnStateClient.handle(requestBuilder, response);
+      return;
+    }
+
+    // Check if data is in the cache already
+    StateGetResponse cachedFirstPage = null;
+    Pair<StateKey, ByteString> cacheKey = new ImmutablePair<>(stateKey, cacheToken);
+    cachedFirstPage = stateCache.getIfPresent(cacheKey);
+
+    switch (request.getRequestCase()) {
+      case GET:
+        if (ByteString.EMPTY.equals(request.getGet().getContinuationToken())) {
+          if (cachedFirstPage == null) {
+            beamFnStateClient.handle(requestBuilder, response);
+            try {
+              stateCache.put(cacheKey, response.get().getGet());
+            } catch (Exception e) {
+              // response should already be completed exceptionally
+              return;
+            }
+          } else {
+            response.complete(
+                responseBuilder.setId(requestBuilder.getId()).setGet(cachedFirstPage).build());
+          }
+        } else {
+          beamFnStateClient.handle(requestBuilder, response);
+        }
+
+        return;
+
+      case APPEND:
+        if (cachedFirstPage == null) {
+          CompletableFuture<StateResponse> responseFuture = new CompletableFuture<>();
+          beamFnStateClient.handle(
+              StateRequest.newBuilder()
```

Comment:
Until we get the rest working, let us treat APPEND as a cache eviction instead. Handling APPEND is more complicated since we want to merge existing data with the new data without needing to have a "get" call and there are a couple of scenarios (e.g., there is no existing data, the last page is still cached, ...).

Reply:
left append as a regular call to the state client for now